File: src/main/scala/uk/co/automatictester/wiremock/maven/plugin/util/ClasspathUtil.scala
package uk.co.automatictester.wiremock.maven.plugin.util

import java.io.File
import java.net.{MalformedURLException, URI, URL}
import java.util

import org.apache.maven.plugin.MojoExecutionException
import org.apache.maven.plugin.descriptor.PluginDescriptor
import org.codehaus.plexus.classworlds.realm.ClassRealm
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.JavaConverters._

class ClasspathUtil {

    var descriptor: PluginDescriptor = _
    var classpathElements: util.List[String] = _
    private val log: Logger = LoggerFactory.getLogger(this.getClass)

    def addRuntimeDependenciesToClasspath(): Unit = {
        val realm: ClassRealm = descriptor.getClassRealm

        toIterator(classpathElements).foreach(classpathElement => {
            log.debug("Adding {} to wiremock-maven-plugin classpath", classpathElement)
            val classpathElementUrl: URL = getClasspathElementFrom(classpathElement)
            realm.addURL(classpathElementUrl)
        })
    }

    private def toIterator[T](javaList: util.List[T]): Iterator[T] = {
        javaList.iterator().asScala
    }

    private def getClasspathElementFrom(classpathElement: String): URL = {
        val classpathElementFile: File = new File(classpathElement)
        val classpathElementUri: URI = classpathElementFile.toURI
        getClasspathElementFrom(classpathElementUri)
    }

    private def getClasspathElementFrom(uri: URI): URL = try {
        uri.toURL
    } catch {
        case e: MalformedURLException => throw new MojoExecutionException("Malformed classpath URL", e)
    }
}

File: src/main/java/uk/co/automatictester/wiremock/maven/plugin/util/ClasspathUtil.java
package uk.co.automatictester.wiremock.maven.plugin.util;

import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugin.descriptor.PluginDescriptor;
import org.codehaus.plexus.classworlds.realm.ClassRealm;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.util.List;

public class ClasspathUtil {

    private PluginDescriptor descriptor;
    private List<String> classpathElements;
    private Logger log = LoggerFactory.getLogger(this.getClass());

    public void setDescriptor(PluginDescriptor pluginDescriptor) {
        this.descriptor = pluginDescriptor;
    }

    public void setClasspathElements(List<String> classpathElements) {
        this.classpathElements = classpathElements;
    }

    public void addRuntimeDependenciesToClasspath() throws MojoExecutionException {
        ClassRealm realm = descriptor.getClassRealm();

        for (String classpathElement : classpathElements) {
            log.debug("Adding {} to wiremock-maven-plugin classpath", classpathElement);
            URL classpathElementUrl = getClasspathElementFrom(classpathElement);
            realm.addURL(classpathElementUrl);
        }
    }

    private URL getClasspathElementFrom(String classpathElement) throws MojoExecutionException {
        File classpathElementFile = new File(classpathElement);
        URI classpathElementUri = classpathElementFile.toURI();
        return getClasspathElementFrom(classpathElementUri);
    }

    private URL getClasspathElementFrom(URI uri) throws MojoExecutionException {
        try {
            return uri.toURL();
        } catch (MalformedURLException e) {
            throw new MojoExecutionException("Malformed classpath URL", e);
        }
    }
}
